scripts/generate_states.py

```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/env python3
+
+"""Generate library/states.c
+states.c is used to output string value of states.
+
+The code generated by this module is only meant to be used in the context
+of that program.
+
+"""
+
+# Copyright The Mbed TLS Contributors
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import sys
+from mbedtls_dev import build_tree
+
+OUTPUT_TEMPLATE = '''\
+/* Automatically generated by generate_states.py. DO NOT EDIT. */
+
+#include "common.h"
+
+#if defined(MBEDTLS_DEBUG_C)
+
+#include "mbedtls/ssl.h"
+#include "ssl_misc.h"
+
+const char *mbedtls_ssl_str_state( mbedtls_ssl_states state )
+{{
+    const char * state_to_str[]=
+    {{
+        {state_list}
+    }};
+
+    if( state > ( sizeof( state_to_str )/sizeof( state_to_str[0]) - 1 ) ||
+        state_to_str[ state ] == NULL )
+    {{
+        return "UNKOWN_STATE";
+    }}
+    return state_to_str[ state ];
+}}
+#endif /* MBEDTLS_DEBUG_C */
+/* End of automatically generated file. */
+
+'''
+def get_enum_definition(name, header='include/mbedtls/ssl.h'):
+    '''
+    Find enum definition in header file
+    '''
+    with open(header, 'r') as f:
+        print_flag = False
+        _temp = ''
+        for line in f:
+            line = line.rstrip()
+            if 'typedef enum' in line:
```

Comment:
Rather than anywhere in the line, which could be in a comment, we should insist that it's at the beginning of the line. Also `enum` should be at the a word, so as not to catch `typedef enumerated_type foo_t;`.

Use regexes, they're typically a cheap way to do reasonably robust parsing, compared to a mere string search which tends to be fragile.

Reply:
moved to #4981